### ISSUE TITLE
Copy management

### ISSUE THREAD

**Issue by tamascz** [textbrowser/biblioteq#83]
Consider the following example:
- A document with 3 copies.
- Copy number 2 with barcode 1-2 reserved, but lost and never will be returned.
- Copy numbers 1 and 3 with barcodes 1-1 and 1-3 reserved and expected to be returned in the future.

Currently I can only "return" the lost copy, but can not delete it or decrease the number of copies (because of reservation) so copies (quantity, total reserved) remains incorrect even if I reserve the lost copy to a fictive ("N/A") member.

Copy Number; Barcode; Availability
1; 1-1; 0
2; 1-2; 1
3; 1-3; 0

The correct method might be (with copies = 2):

1; 1-1; 0
2; "DELETED"
3; 1-3; 0

or maybe

1; 1-1; 0
2; 1-3; 0

1; 1-1; 0
3; 1-3; 0

Reservation Status window is very small and not retain manual resize.

**Comment by textbrowser** [comment]
It's you! :) Thanks for the ticket. Reviewing.

**Comment by textbrowser** [comment]
Aha, this be an enhancement.

**Comment by textbrowser** [comment]
I am relearning to type so the changes will be sloooooooooooooooooooooooooooooooooooooooooooooooo^100w. OK?

**Comment by textbrowser** [comment]
The previous sentences required 32 seconds to complete.

**Comment by textbrowser** [comment]
A new status field is required. Deleted, Available, Borrowed, Lost.

**Comment by textbrowser** [comment]
As for the size of the panel, I will increase its default size.

**Comment by textbrowser** [comment]
Default size increased. Added new status fields to database tables. SQLite-upgrade process remains. Then, database and interface modifications. Database changes require significant efforts.

**Comment by textbrowser** [comment]
Thoughts?

![Screenshot from 2020-12-10 19-02-27](https://user-images.githubusercontent.com/10701156/101844416-5bc8b200-3b1a-11eb-8165-66adb946b67b.png)

**Comment by textbrowser** [comment]
Yes, the quantity and the status are independent.

**Comment by tamascz** [comment]
So you want to count lost and deleted too in the actual number of copies (quantity):
2 borrowed, 1 available = 3 copies (quantity)
1 lost, 1 deleted, 1 available = 3 copies (instead of 1 copy)
or the other way (3, 1)?

**Comment by textbrowser** [comment]
The quantity represents the physical number of copies that a particular item had or has. If a copy's status is set to deleted or lost, the quantity is not automatically changed. Only an administrator may decrease or increase the quantity. If the quantity is decreased such that the new quantity will not harm borrowed items, copies are removed from the database.

The new status field does not impose logical significance, that is, the field is a visual reminder of a copy.

Current values of status: available, deleted, lost. An item may have had 5 copies, 3 of which are marked as deleted. An administrator may decide to leave the quantity at 5 for historic reasons. Or, they may modify the quantity to 2 if the "deleted" items are not borrowed. Recall that an artificial process of "returning" a copy may allow "deleted" copies to be removed from the database.

In conclusion, status is a visual guide and BQ provides the mechanisms for maintaining historic perspectives as well as maintaining records mirroring physical artifacts.

Hope this is clear. :)

**Comment by textbrowser** [comment]
Also, copy statuses are visible during the reservation process. The belief is that the administrator is careful to reserve the correct copy.

**Comment by tamascz** [comment]
Clear and seems flexible.
So instead of this:
> A new status field is required. Deleted, Available, Borrowed, Lost.
you decided that borrowed will not be a value of the status field, and to keep managing reservations in the current way  (or some other way, but independently from the new status field).
And lost is another possible "label" or "reminder" value of the status field, which allows the removal of the actual copy record if the quantity decreased, just like the value "deleted".

**Comment by textbrowser** [comment]
Yah, I considered borrowed as a value but this would require a relationship between the status and the reservation process. Future people may want more status values and a relationship could become more nebulous.

It's like this: we have a bunch of copies and some of those copies were lost and some were deleted but this is the original quantity we had and we'd like to portray this history. Or, this quantity actually represents the number of physical copies we have and we've labeled them as being available. So, both situations are possible and BQ delegates the responsibility onto the owners of the data. More logic sometimes decreases flexibility.

**Comment by textbrowser** [comment]
BQ is almost 15 years old and it's quite mature and even little database changes require significant effort. The belief that defining a new item or a new field is simple plug-and-play is just not true. There's also the additional work because of two database engines. PQ and SQLite are not totally compatible. People also expect that in addition to creating new SQLite schemas (whenever you create a new file), they have a tool to upgrade their existing database files. Luckily, PQ expects more hands-on people and less GUI-GUI people. :)

**Comment by textbrowser** [comment]
Scheduled for 2020.01.01. There are some other reminder items and perhaps at least one more will be completed. So, closing!

**Comment by textbrowser** [comment]
Like: damaged, sold, transferred, unknown. Or, descriptive text.

**Comment by textbrowser** [comment]
Actually, 16 years old.

**Comment by tamascz** [comment]
Yes, more feasible, development friendly and future proof this way. (And my reasons for choosing just the easier SQLite implementation over PSQL are the small, one person school library.)

**Comment by textbrowser** [comment]
PQ is the way to go for any use. Network.

**Comment by textbrowser** [comment]
2020.01.01 release is postponed. Sorry!

**Comment by tamascz** [comment]
Thanks for the new status field.

But the original intention of my request was reflecting the actual, real copies in the quantity/availability numbers with regard to deleted/lost copies (because I cannot "delete" in the middle of the copies, only decrease the number of copies from the end).

Test case:

- Add a book with 3 copies.
- Imagine that 2 copies returned, but copy number 2 never will be returned, so change it's status to deleted/lost (possibly "return" it).
- Search for the book in the main window, quantity is still 3 not 2 (availability still 2 / or 3, and total reserved is 1 / or 0).

**Comment by textbrowser** [comment]
Released!

**Comment by textbrowser** [comment]
Where did the comment go? Deleted?

**Comment by tamascz** [comment]
Sorry, I not remembered exactly that the new field eventually will be only an attribute, not a solution to my problem (my concentration and memory are too compromised).

Either I struggle along with the barcode swapping (and then decreasing copies ) or find some other option for the catalog than BQ.

**Comment by textbrowser** [comment]
![Screenshot_20210204_090112](https://user-images.githubusercontent.com/10701156/106903120-99cff580-66c7-11eb-9eaa-f79d534610c6.png)

The new implementation is simple. In the example, I have 5 copies of a book. 4 copies are not physically available. I retain their information without having to decrease the number of copies.

So, an additional change will be to include a removal of an copy, decreasing the copy number, and correctly adjusting the display. OK!

**Comment by tamascz** [comment]
Yes, this would be a final solution to my copy managing problems: removing (or inactivating?) copies from the middle, showing the actual/real number of copies under quantity, availability and total reserved.

**Comment by textbrowser** [comment]
Will complete this by the middle of February.

**Comment by textbrowser** [comment]
The original design also caused the misalignment of reserved copy numbers.

**Comment by textbrowser** [comment]
All changes are now committed. Which operating system do you use?

**Comment by textbrowser** [comment]
Release in four days. May change.

**Comment by textbrowser** [comment]
The original problem also involved borrowed items. For example, suppose copy number 3 was borrowed and copy number 2 was deleted. Copy number 3 would become copy number 2 and the original (#3) would not be visible as a copy that was borrowed. Corrected.

**Comment by tamascz** [comment]
> The original problem also involved borrowed items. For example, suppose copy number 3 was borrowed and copy number 2 was deleted. Copy number 3 would become copy number 2 and the original (#3) would not be visible as a copy that was borrowed. Corrected.

Now that you mention it, exactly (when delete button was in copy browser).

OS is Windows. Thanks for the fixes!

**Comment by textbrowser** [comment]
Released! Please test thoroughly and report problems.
